File: APAC Semifinal/millionaire.py
# Copyright (c) 2018 kamyu. All rights reserved.
#
# Google Code Jam 2008 APAC Semifinal - Problem C. Millionaire
# https://codejam.withgoogle.com/codejam/contest/32005/dashboard#s=p2
#
# Time:  O(4^M)
# Space: O(2^M)
#

TARGET = 1000000
MAX_M = 15
dp = [[0.0 for _ in xrange(2**MAX_M+1)] for _ in xrange(2)]

def millionaire():
    M, P, X = raw_input().strip().split()
    M, P, X = int(M), float(P), int(X)
    dp[0][1] = 1.0
    for m in xrange(1, M+1):
        n = 2**m
        curRound, nxtRound = dp[m%2], dp[(m-1)%2]
        for sum in xrange(n+1):
            max_bet = min(sum, n-sum)
            curRound[sum] = nxtRound[sum//2];
            for bet in xrange(1, max_bet+1):
                curRound[sum] = max(curRound[sum],
                                    P*nxtRound[(sum+bet)//2] + (1.0-P)*nxtRound[(sum-bet)//2])
    return dp[M%2][X*(2**M)//TARGET]

for case in xrange(input()):
    print 'Case #%d: %s' % (case+1, millionaire())

File: APAC Semifinal/millionaire.cpp
// Copyright (c) 2018 kamyu. All rights reserved.

/*
 * Google Code Jam 2008 APAC Semifinal - Problem C. Millionaire
 * https://codejam.withgoogle.com/codejam/contest/32005/dashboard#s=p2
 *
 * Time:  O(4^M)
 * Space: O(2^M)
 *
 */

#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::endl;
using std::vector;
using std::max;
using std::min;

const int TARGET = 1000000;
const int MAX_X = 15;
double dp[2][(1 << MAX_X) + 1];

double millionaire() {
    uint64_t M, X;
    double P;
    cin >> M >> P >> X;
    dp[0][1] = 1.0;
    for (int m = 1; m <= M; ++m) {
        const auto n = 1 << m;
        auto& curRound = dp[m % 2];
        auto& nxtRound = dp[(m - 1) % 2];
        for (int sum = 0; sum <= n; ++sum) {
            curRound[sum] = nxtRound[sum / 2];
            const auto max_bet = min(sum, n - sum);
            for (int bet = 1; bet <= max_bet; ++bet) {
                const auto tmp = P * nxtRound[(sum + bet) / 2] +
                                 (1 - P) * nxtRound[(sum - bet) / 2];
                if (curRound[sum] < tmp) {
                    curRound[sum] = tmp;
                }
            }
        }
    }
    return dp[M % 2][X * (1 << M) / TARGET];
}

int main() {
    int T;
    cin >> T;
    for (int test = 1; test <= T; ++test) {
        cout << "Case #" << test << ": " << millionaire() << endl;
    }
    return 0;
}
